samples/immediate-triangle/main.c

```diff
@@ -0,0 +1,187 @@
+/*
+ * This sample provides a very basic demonstration of 3D rendering on the Xbox,
+ * using pbkit. Based on the pbkit demo sources.
+ */
+#include <hal/video.h>
+#include <hal/xbox.h>
+#include <math.h>
+#include <pbkit/pbkit.h>
+#include <xgu/xgu.h>
+#include <xgu/xgux.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <strings.h>
+#include <xboxkrnl/xboxkrnl.h>
+#include <hal/debug.h>
+#include <windows.h>
+
+
+/* Main program function */
+int main(void)
+{
+    uint32_t *p;
+
+    XVideoSetMode(640, 480, 32, REFRESH_DEFAULT);
+
+    int status = pb_init();
+    if (status) {
+        debugPrint("pb_init Error %d\n", status);
+        Sleep(2000);
+        XReboot();
+        return 1;
+    }
+
+    //FIXME: Hack for XQEMU
+    Sleep(1000);
+
+    pb_show_front_screen();
+
+    /* Basic setup */
+    int width = pb_back_buffer_width();
+    int height = pb_back_buffer_height();
+
+    /* A generic identity matrix */
+    const float m_identity[4*4] = {
+        1.0f, 0.0f, 0.0f, 0.0f,
+        0.0f, 1.0f, 0.0f, 0.0f,
+        0.0f, 0.0f, 1.0f, 0.0f,
+        0.0f, 0.0f, 0.0f, 1.0f
+    };
+
+    /* A viewport matrix; this maps:
+     *  - X from [-1, 1] to [0, width]
+     *  - Y from [-1, 1] to [height, 0]
+     *  - Z from [ 0, 1] to [0, 0xFFFF]
+     *  - W to 1
+     * This scales: X from [-1, 1] to [0, width]
+     */
+    float m_viewport[4*4] = {
+        width/2.0f, 0.0f,         0.0f,          width/2.0f,
+        0.0f,       height/-2.0f, 0.0f,          height/2.0f,
+        0.0f,       0.0f,         (float)0xFFFF, 0.0f,
+        0.0f,       0.0f,         0.0f,          1.0f
+    };
+
+    /* Setup fragment shader */
+    p = pb_begin();
+    //FIXME: The MASK macro should be contained within ps.inl
+    //FIXME: Ideally ps.inl would contain a struct, not this weird garbage
+    #pragma push_macro("MASK")
+    #undef MASK //FIXME: Maybe not necessary?
+    #define MASK(mask, val) (((val) << (__builtin_ffs(mask)-1)) & (mask))
+    #include "ps.inl"
+    #undef MASK //FIXME: Maybe not necessary?
+    #pragma pop_macro("MASK")
+    pb_end(p);
+
+    /* Set up some default GPU state (should be done in xgux_init maybe? currently partially done in pb_init) */
+    {
+        p = pb_begin();
+
+        //FIXME: p = xgu_set_skinning(p, XGU_SKINNING_OFF);
+        //FIXME: p = xgu_set_normalization(p, false);
+        p = xgu_set_lighting_enable(p, false);
+
+        for(int i = 0; i < XGU_TEXTURE_COUNT; i++) {
+            //FIXME: p = xgu_set_texgen(p, XGU_TEXGEN_OFF);
+            //p = xgu_set_texture_matrix_enable(p, i, false);
```

Comment:
This can be this now:

```c
  for(int i = 0; i < XGU_TEXTURE_COUNT; i++) {
      p = xgu_set_texgen_s(p, i, XGU_TEXGEN_DISABLE);
      p = xgu_set_texgen_t(p, i, XGU_TEXGEN_DISABLE);
      p = xgu_set_texgen_r(p, i, XGU_TEXGEN_DISABLE);
      p = xgu_set_texgen_q(p, i, XGU_TEXGEN_DISABLE);
      p = xgu_set_texture_matrix_enable(p, i, false);
      p = xgu_set_texture_matrix_enable(p, i, m_identity);
  }
```